src/go/util/httppattern/uri_template_test.go

```diff
@@ -0,0 +1,1008 @@
+      "StartSegment": 0
+    }
+  ],
+  "Verb": "myverb"
+}
+		`},
+		{
+			desc:        "VariableAndCustomVerbTests-6",
+			UriTemplate: "/{x.y.z=a/**/b}:custom",
+			wantUriTemplate: `
+{
+  "Segments": [
+    "a",
+    "**",
+    "b"
+  ],
+  "Variables": [
+    {
+      "EndSegment": -2,
+      "FieldPath": [
+        "x",
+        "y",
+        "z"
+      ],
+      "HasWildcardPath": true,
+      "StartSegment": 0
+    }
+  ],
+  "Verb": "custom"
+}
+		`},
+		{
+			desc:        "VariableAndCustomVerbTests-7",
+			UriTemplate: "/{x.y.z=a/**/b}/c/d:custom",
+			wantUriTemplate: `
+{
+  "Segments": [
+    "a",
+    "**",
+    "b",
+    "c",
+    "d"
+  ],
+  "Variables": [
+    {
+      "EndSegment": -4,
+      "FieldPath": [
+        "x",
+        "y",
+        "z"
+      ],
+      "HasWildcardPath": true,
+      "StartSegment": 0
+    }
+  ],
+  "Verb": "custom"
+}
+		`},
+		{
+			desc:        "RootPath",
+			UriTemplate: "/",
+			wantUriTemplate: `
+{
+  "Segments": null,
+  "Variables": null,
+  "Verb": ""
+}
+		`},
+		{
+			desc:        "ParseVerbTest2",
+			UriTemplate: "/a/*:verb",
+			wantUriTemplate: `
+{
+  "Segments": [
+    "a",
+    "*"
+  ],
+  "Variables": null,
+  "Verb": "verb"
+}
+		`},
+		{
+			desc:        "ParseVerbTest3",
+			UriTemplate: "/a/**:verb",
+			wantUriTemplate: `
+{
+  "Segments": [
+    "a",
+    "**"
+  ],
+  "Variables": null,
+  "Verb": "verb"
+}
+		`},
+		{
+			desc:        "ParseVerbTest4",
+			UriTemplate: "/a/{b=*}/**:verb",
+			wantUriTemplate: `
+{
+  "Segments": [
+    "a",
+    "*",
+    "**"
+  ],
+  "Variables": [
+    {
+      "EndSegment": 2,
+      "FieldPath": [
+        "b"
+      ],
+      "HasWildcardPath": false,
+      "StartSegment": 1
+    }
+  ],
+  "Verb": "verb"
+}
+		`},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			ht := Parse(tc.UriTemplate)
+			if ht == nil {
+				t.Fatal("fail to generate UriTemplate")
+			}
+
+			if tc.wantUriTemplate != "" {
+				bytes, _ := json.Marshal(ht)
+				getUriTemplate := string(bytes)
+
+				if err := util.JsonEqual(tc.wantUriTemplate, getUriTemplate); err != nil {
+					t.Error(err)
+				}
+			}
+		})
+	}
+}
+
+func TestUriTemplateParseError(t *testing.T) {
+	testCases := []string{
+		"",
+		"//",
+		"/{}",
+		"/a/",
+		"/a//b",
+		":verb",
+		"/:verb",
+		"/a/:verb",
+		":",
+		"/:",
+		"/*:",
+		"/**:",
+		"/{{",
+		"/{var}:",
+		"/{var}::",
+		"/{var/a",
+		"/{{var}}",
+		"/a/b/:",
+		"/a/b/*:",
+		"/a/b/**:",
+		"/a/b/{var}:",
+		"/a/{",
+		"/a/{var",
+		"/a/{var.",
+		"/a/{x=var:verb}",
+		"a",
+		"{x}",
+		"{x=/a}",
+		"{x=/a/b}",
+		"a/b",
+		"a/b/{x}",
+		"a/{x}/b",
+		"a/{x}/b:verb",
+		"/a/{var=/b}",
+		"/{var=a/{nested=b}}",
+		"/a{x}",
+		"/{x}a",
+		"/a{x}b",
+		"/{x}a{y}",
+		"/a/b{x}",
+		"/a/{x}b",
+		"/a/b{x}c",
+		"/a/{x}b{y}",
+		"/a/b{x}/s",
+		"/a/{x}b/s",
+		"/a/b{x}c/s",
+		"/a/{x}b{y}/s",
+		// No verb test cases.
+		":",
+		"/:",
+		"/a/:",
+		"/a/*:",
+		"/a/**:",
+		"/a/{b=*}/**:",
+	}
```

Comment:
A few more cases I'd like to verify:
- /a/**/b/*/c

Reply:
Done.